File: android-native/app/src/main/java/flexr/social/app/data/repository/MessageRepository.kt
package flexr.social.app.data.repository

import flexr.social.app.core.network.apiCall
import flexr.social.app.data.local.MessageDao
import flexr.social.app.data.local.MessageEntity
import flexr.social.app.data.local.toDomain
import flexr.social.app.data.local.toEntity
import flexr.social.app.data.remote.FlexrApi
import flexr.social.app.data.remote.dto.SendMessageRequestDto
import flexr.social.app.domain.model.Message
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import java.time.Instant
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Chatverlauf eines Matches.
 *
 * Room ist die Anzeigequelle, das Netz der Lieferant. Gesendete Nachrichten
 * erscheinen sofort als „pending" in der Liste und werden durch die
 * Serverantwort ersetzt — das Web-Frontend musste dafür neu laden.
 */
@Singleton
class MessageRepository @Inject constructor(
    private val api: FlexrApi,
    private val messageDao: MessageDao,
) {

    fun messages(matchId: String): Flow<List<Message>> =
        messageDao.observeForMatch(matchId).map { rows -> rows.map { it.toDomain() } }

    /**
     * Holt den Verlauf. Der Aufruf markiert serverseitig zugleich alle
     * Nachrichten der Gegenseite als gelesen (siehe messages.py).
     */
    suspend fun refresh(matchId: String) {
        val remote = apiCall { api.getMessages(matchId) }.map { it.toDomain() }
        messageDao.replaceSynced(matchId, remote.map { it.toEntity() })
    }

    suspend fun send(matchId: String, senderId: String, content: String): Message {
        val pendingId = "pending-${UUID.randomUUID()}"
        val pending = MessageEntity(
            id = pendingId,
            matchId = matchId,
            senderId = senderId,
            content = content,
            createdAt = Instant.now(),
            readAt = null,
            wasCensored = false,
            isPending = true,
        )
        messageDao.upsert(pending)

        return try {
            val sent = apiCall {
                api.sendMessage(matchId, SendMessageRequestDto(content))
            }.toDomain()
            messageDao.deleteById(pendingId)
            messageDao.upsert(sent.toEntity())
            sent
        } catch (throwable: Throwable) {
            // Fehlgeschlagene Nachricht nicht als „zugestellt" stehen lassen —
            // der Eingabetext wird im ViewModel wiederhergestellt.
            messageDao.deleteById(pendingId)
            throw throwable
        }
    }

    /**
     * Chatverlauf leeren — wirkt nur für die leerende Seite. Serverseitig wird
     * lediglich ein „geleert ab"-Zeitpunkt gesetzt, für die andere Person
     * bleibt der Verlauf erhalten.
     */
    suspend fun clearHistory(matchId: String) {
        apiCall { api.clearMessages(matchId) }
        messageDao.deleteForMatch(matchId)
    }
}

File: ios/FLEXR/Data/Repository/MessageRepository.swift
import Foundation

/// Chatverlauf eines Matches.
///
/// Der lokale Bestand ist die Anzeigequelle, das Netz der Lieferant. Gesendete
/// Nachrichten erscheinen sofort als „pending" in der Liste und werden durch die
/// Serverantwort ersetzt — das Web-Frontend musste dafür neu laden.
@MainActor
final class MessageRepository {

    private let api: FlexrAPI
    private let store: FlexrStore

    init(api: FlexrAPI, store: FlexrStore) {
        self.api = api
        self.store = store
    }

    func messages(matchID: String) -> [Message] {
        store.messages(matchID: matchID).map { $0.toDomain() }
    }

    /// Holt den Verlauf. Der Aufruf markiert serverseitig zugleich alle
    /// Nachrichten der Gegenseite als gelesen (siehe messages.py).
    func refresh(matchID: String) async throws {
        let remote = try await api.messages(matchID: matchID).map { $0.toDomain() }
        store.replaceSyncedMessages(matchID: matchID, with: remote)
    }

    /// Legt die Zeile sofort örtlich an und liefert ihre Kennung.
    ///
    /// Bewusst getrennt von [send] und nicht `async`: Die eigene Nachricht soll
    /// im selben Bild erscheinen, in dem der Knopf gedrückt wurde — nicht erst
    /// nach der Antwort des Servers. Vorher steckte das Einfügen im
    /// Netzaufruf, und die Ansicht lud ihre Liste davor neu: Die Zeile war
    /// bis zur Serverantwort unsichtbar.
    func insertPending(matchID: String, senderID: String, content: String) -> String {
        let pendingID = "pending-\(UUID().uuidString)"
        let pending = Message(
            id: pendingID,
            matchID: matchID,
            senderID: senderID,
            content: content,
            createdAt: Date(),
            readAt: nil,
            wasCensored: false
        )
        store.insert(pending, isPending: true)
        return pendingID
    }

    /// Schickt eine mit [insertPending] angelegte Zeile ab und ersetzt sie
    /// durch die Fassung des Servers (der den Text zensiert haben kann).
    @discardableResult
    func send(matchID: String, pendingID: String, content: String) async throws -> Message {
        do {
            let sent = try await api.sendMessage(
                matchID: matchID,
                body: SendMessageRequestDTO(content: content)
            ).toDomain()
            store.deleteMessage(id: pendingID)
            store.insert(sent, isPending: false)
            return sent
        } catch {
            // Fehlgeschlagene Nachricht nicht als „zugestellt" stehen lassen —
            // der Eingabetext wird im ViewModel wiederhergestellt.
            store.deleteMessage(id: pendingID)
            throw error
        }
    }

    /// Chatverlauf leeren — wirkt nur für die leerende Seite. Serverseitig wird
    /// lediglich ein „geleert ab"-Zeitpunkt gesetzt, für die andere Person
    /// bleibt der Verlauf erhalten.
    func clearHistory(matchID: String) async throws {
        try await api.clearMessages(matchID: matchID)
        store.deleteMessages(matchID: matchID)
    }
}
